### PR TITLE
python3Packages.hy: 1.0a3 -> 1.0a4; hy: replace it with python3Packages.hy; hyrule: init at 0.1

### LINKED ISSUE #171428: We have two separate Hy derivations
### Describe the bug

```
~/Projects/nixpkgs master
❯ rg "\bhy ="
pkgs/top-level/python-packages.nix
3952:  hy = callPackage ../development/python-modules/hy { };

pkgs/top-level/all-packages.nix
35035:  hy = callPackage ../development/interpreters/hy {};

pkgs/development/libraries/aspell/dictionaries.nix # Ignore this, another thing not related
479:  hy = buildOfficialDict {
```

The one in `python3Packages` is more up-to-date (1.0a3 vs 1.0a1, current version is 1.0a4 though), ~~however it seems also more incomplete (it is missing some dependencies, like `fastentrypoints` that reduces interpreter startup)~~. The second one is the oldest though.

Not sure what is the best solution here. It doesn't seem to make much sense to have Hy being considered a interpreter instead of a Python package (it even has `hyDefinedPythonPackages` attribute that AFAIK isn't any different from simply from using `buildPythonPackages` with `hy` and other dependencies as `buildInputs`). But I don't have sufficient context here.

We probably need to merge one in another and delete the other derivation.

### Steps To Reproduce
1. `cd nixpkgs && nix repl`
2. `:lf ./`
3. ```
   nix-repl> legacyPackages.x86_64-linux.hy == legacyPackages.x86_64-linux.python3Packages.hy
   false # should be true
   ```
### Expected behavior
We should only have one Hy package.

### Screenshots
N/A

### Additional context
N/A

### Notify maintainers
<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->
@fabaff (`python3Packages.hy`)
@nixy (`hy`)
@Mazurel (`hy`)

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.36, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4`
 - nixpkgs: `/nix/store/fj0r6l3gglr0k0b4dxn3hm8drcwclm36-source`
```

### PR DESCRIPTION
**Pull request by thiagokokada**
###### Description of changes

Fixes: https://github.com/NixOS/nixpkgs/issues/171428

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by thiagokokada** [comment]
CC @fabaff @nixy @Mazurel

**Review by fabaff — COMMENTED** [review]
Looks like that I added a "duplicate" at some point. Thanks for cleaning it up.

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 171432` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>hy (python39Packages.hy)</li>
    <li>python310Packages.hy</li>
  </ul>
</details>

**Comment by thiagokokada** [comment]
Huh... I think the GitHub source doesn't have the version attribute set:

```
Hy unknown using CPython(main) 3.9.12 on Linux
=> (+ 1 1)
2
=>
```

Compares this to `interpreters/hy`:

```
hy 1.0a1 using CPython(main) 3.9.12 on Linux
=>
```

Let me see if I can patch this.

**Comment by thiagokokada** [comment]
> Let me see if I can patch this.

Done. I also added a test to ensure that the version is being set correctly.

**Comment by thiagokokada** [comment]
@ofborg eval

**Comment by thiagokokada** [comment]
With commit [24c9621](https://github.com/NixOS/nixpkgs/pull/171432/commits/24c96215cc7902dc72ea712ac8e4c64b545bf126) this works:

```
Welcome to Nix 2.4. Type :? for help.

nix-repl> :lf ./
Added 16 variables.

nix-repl> legacyPackages.x86_64-linux.hy.withPackages (ps: with ps; [ pytest ])
«derivation /nix/store/748nk43sfswh1i8d9jjl0rb647j7hx0f-python3.9-hy-1.0a4.drv»
```

In `master`:

```
❯ nix repl
Welcome to Nix 2.4. Type :? for help.

nix-repl> :lf ./
Added 16 variables.

nix-repl> legacyPackages.x86_64-linux.hy.withPackages (ps: with ps; [ pytest ])
«derivation /nix/store/62b9124kh079v3jr1sngh7hglyz6cfrw-hy-1.0a1.drv»
```

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 171432` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>hy (python39Packages.hy)</li>
    <li>python310Packages.hy</li>
  </ul>
</details>

**Comment by thiagokokada** [comment]
I added `hyrule` too, since on version 1.0a4 most of Hy stdlib was moved to this library.

**Comment by WolfangAukang** [comment]
Result of `nixpkgs-review pr 171432` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>hy (python39Packages.hy)</li>
    <li>python310Packages.hy</li>
    <li>python310Packages.hyrule</li>
    <li>python39Packages.hyrule</li>
  </ul>
</details>

File: pkgs/development/python-modules/hy/default.nix [hunk 1]
```diff
@@ -4,31 +4,43 @@
...
 , rply
+, testers
+, toPythonApplication
+, hyDefinedPythonPackages ? python-packages: [ ] /* Packages like with python.withPackages */
 }:
 
 buildPythonPackage rec {
   pname = "hy";
-  version = "1.0a3";
+  version = "1.0a4";
```
Comment (WolfangAukang) [inline pkgs/development/python-modules/hy/default.nix:20]:
```suggestion
format = "setuptools";

```

A minor nitpick, but would be helpful for maintenance.

File: pkgs/development/python-modules/hyrule/default.nix [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+, fetchFromGitHub
+, hy
+, pytestCheckHook
+, pythonOlder
+}:
+
+buildPythonPackage rec {
+  pname = "hyrule";
+  version = "0.1";
+
```
Comment (WolfangAukang) [inline pkgs/development/python-modules/hyrule/default.nix:12]:
```suggestion
format = "setuptools";

```

Same as from above

**Review by AtilaSaraiva — APPROVED** [review]
Result of `nixpkgs-review pr 171432` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>hy (python39Packages.hy)</li>
    <li>python310Packages.hy</li>
    <li>python310Packages.hyrule</li>
    <li>python39Packages.hyrule</li>
  </ul>
</details>

**Comment by AtilaSaraiva** [comment]
I don't understand much of what hy does but it makes sense to do this change.

**Review by WolfangAukang — APPROVED** [review]
LGTM

**Review by ratsclub — APPROVED** [review]

**Pull request merged**